### PR TITLE
Implementation of Label Definition update

### PR DESCRIPTION
**Pull request by kfurgol**
**Description**

Changes proposed in this pull request:

- Implemented Label Definition update

**Related issue(s)**
#154

### CODE REVIEW

File: components/director/internal/domain/labeldef/service.go [hunk 1]
```diff
@@ -43,3 +46,29 @@ func (s *service) Create(ctx context.Context, def model.LabelDefinition) (model.
...
+	ld, err := s.repo.GetByKey(ctx, def.Tenant, def.Key)
+	if err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while receiving Label Definition")
+	}
+
+	id := ld.ID
+	ld = &def
+	ld.ID = id
+
+	if err := s.repo.Update(ctx, *ld); err != nil {
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service.go]: what about checking if existing labels are valid against the new definition?

File: components/director/internal/domain/labeldef/service.go [hunk 2]
```diff
@@ -43,3 +46,29 @@ func (s *service) Create(ctx context.Context, def model.LabelDefinition) (model.
...
+	if err := def.ValidateForUpdate(); err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while validating Label Definition")
+	}
+
+	ld, err := s.repo.GetByKey(ctx, def.Tenant, def.Key)
+	if err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while receiving Label Definition")
+	}
+
+	id := ld.ID
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: added check

File: components/director/internal/domain/labeldef/service.go (hunk 1 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/model/labeldef.go [hunk 3]
```diff
@@ -32,3 +32,20 @@ func (def *LabelDefinition) Validate() error {
 
 	return nil
 }
+
+func (def *LabelDefinition) ValidateForUpdate() error {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/model/labeldef.go]: done

File: components/director/internal/domain/labeldef/service.go [hunk 4]
```diff
@@ -60,3 +75,52 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+	if err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while receiving Label Definition")
+	}
+
+	if ld.ID == "" {
+		return model.LabelDefinition{}, fmt.Errorf("id of received label definition is empty")
+	}
+
+	id := ld.ID
+	ld = &def
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/service.go]: Do we really want to override all the properties, or should we only allow to update schema? For example I would not update the tenant property.

File: components/director/internal/domain/labeldef/service.go [hunk 5]
```diff
@@ -60,3 +75,52 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+			return model.LabelDefinition{}, fmt.Errorf("label with key %s is not valid against new schema", label.Key)
+		}
+
+	}
+
+	if err := s.repo.Update(ctx, *ld); err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while updating Label Definition")
+	}
+
+	updatedLabelDef, err := s.repo.GetByKey(ctx, ld.Tenant, ld.Key)
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/service.go]: Is this call really necessary? Can't we just return the object that was just used in the repository Update call?

File: components/director/internal/domain/labeldef/service_test.go [hunk 6]
```diff
@@ -140,19 +141,288 @@ func TestServiceList(t *testing.T) {
...
+		// GIVEN
+
+		sut := labeldef.NewService(nil, nil, nil)
+		// WHEN
+		_, err := sut.Update(context.TODO(), model.LabelDefinition{})
+		// THEN
+		require.EqualError(t, err, "while validating Label Definition: missing Tenant field")
+	})
+
+	t.Run("returns error if Label Definition was not found", func(t *testing.T) {
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/service_test.go]: This covers the scenario when the repository returns an error. The `GetByKey` method returns `nil, nil` when there is no record.

File: components/director/internal/model/labeldef.go [hunk 7]
```diff
@@ -32,3 +32,20 @@ func (def *LabelDefinition) Validate() error {
...
+	if def.Tenant == "" {
+		return errors.New("missing Tenant field")
+	}
+
+	if def.Key == "" {
+		return errors.New("missing Key field")
+	}
+	if def.Schema != nil {
```
Comment (crabtree) [inline components/director/internal/model/labeldef.go]: New line missing

File: components/director/internal/domain/labeldef/service.go [hunk 8]
```diff
@@ -60,3 +75,52 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+	if err := def.ValidateForUpdate(); err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while validating Label Definition")
+	}
+
+	ld, err := s.repo.GetByKey(ctx, def.Tenant, def.Key)
+	if err != nil {
+		return model.LabelDefinition{}, errors.Wrap(err, "while receiving Label Definition")
+	}
+
+	if ld.ID == "" {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: added check and test case

File: components/director/internal/domain/labeldef/service.go (hunk 4 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: changed to update only schema

File: components/director/internal/domain/labeldef/service.go (hunk 5 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: removed that call

File: components/director/internal/domain/labeldef/service_test.go (hunk 6 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: there are now 2 cases, when error is returned and when nil is returned

File: components/director/internal/model/labeldef.go (hunk 7 above)
Comment (kfurgol) [reply components/director/internal/model/labeldef.go]: done

File: components/director/pkg/jsonschema/validator.go [hunk 9]
```diff
@@ -64,7 +66,11 @@ func (v *validator) ValidateRaw(value interface{}) (bool, error) {
 		return true, nil
 	}
 
-	jsonLoader := gojsonschema.NewRawLoader(value)
+	valueString, err := json.Marshal(value)
+	if err != nil {
+		return false, err
+	}
+	jsonLoader := gojsonschema.NewStringLoader(string(valueString))
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/pkg/jsonschema/validator.go]: done

File: components/director/internal/domain/labeldef/service.go [hunk 10]
```diff
@@ -60,3 +75,45 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+	}
+
+	for _, label := range existingLabels {
+		ok, err := validator.ValidateRaw(label)
+		if err != nil {
+			return model.LabelDefinition{}, errors.Wrap(err, "while validating existing labels against new schema")
+		}
+
+		if ok == false {
+			return model.LabelDefinition{}, fmt.Errorf("label with key %s is not valid against new schema", label.Key)
```
Comment (dbadura) [inline components/director/internal/domain/labeldef/service.go]: You can use `errors` package.

File: components/director/internal/domain/label/repository_test.go [hunk 11]
```diff
@@ -511,7 +511,202 @@ func TestRepository_List(t *testing.T) {
...
+		ctx := context.TODO()
+		ctx = persistence.SaveToContext(ctx, db)
+		// WHEN
+		actual, err := repo.ListByKey(ctx, tnt, labelKey)
+		// THEN
+		require.NoError(t, err)
+		assert.ElementsMatch(t, expected, actual)
+	})
+
+	t.Run("Success - Labels only for Applications", func(t *testing.T) {
```
Comment (aszecowka) [inline components/director/internal/domain/label/repository_test.go]: we have 3 use cases which can be reduced to only 1

File: components/director/internal/domain/labeldef/service_test.go [hunk 12]
```diff
@@ -140,19 +141,232 @@ func TestServiceList(t *testing.T) {
...
+
+		mockLabelRepository.On("ListByKey", context.TODO(), tenant, key).Return(existingLabels, nil).Once()
+
+		ctx := context.TODO()
+		sut := labeldef.NewService(mockRepository, mockLabelRepository, nil)
+		// WHEN
+		_, err := sut.Update(ctx, in)
+		// THEN
+		require.Error(t, err)
+		require.EqualError(t, err, "label with key firstName is not valid against new schema for Runtime with ID foo")
```
Comment (dbadura) [inline components/director/internal/domain/labeldef/service_test.go]: Instead of calling `fixBasicSchema`, call `fixSchema` with property name  `firstName` It's easier to read.

File: components/director/internal/domain/label/repository_test.go [hunk 13]
```diff
@@ -511,7 +511,202 @@ func TestRepository_List(t *testing.T) {
...
+		ctx := context.TODO()
+		ctx = persistence.SaveToContext(ctx, db)
+		// WHEN
+		actual, err := repo.ListByKey(ctx, tnt, labelKey)
+		// THEN
+		require.NoError(t, err)
+		assert.ElementsMatch(t, expected, actual)
+	})
+
+	t.Run("Error - Doesn't exist", func(t *testing.T) {
```
Comment (aszecowka) [inline components/director/internal/domain/label/repository_test.go]: `Error - Doesn't exist`
and 
```
require.NoError(t, err)
assert.Empty(t, actual)
```

make it consitent

File: components/director/internal/domain/labeldef/service.go [hunk 14]
```diff
@@ -60,3 +75,45 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+		return model.LabelDefinition{}, errors.Wrap(err, "while creating validator for new schema")
+	}
+
+	for _, label := range existingLabels {
+		ok, err := validator.ValidateRaw(label)
+		if err != nil {
+			return model.LabelDefinition{}, errors.Wrap(err, "while validating existing labels against new schema")
+		}
+
+		if ok == false {
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service.go]: `if !ok`

File: components/director/internal/domain/labeldef/service.go [hunk 15]
```diff
@@ -60,3 +75,45 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+
+	if ld == nil {
+		return model.LabelDefinition{}, fmt.Errorf("definition with %s key doesn't exist", def.Key)
+	}
+
+	ld.Schema = def.Schema
+
+	existingLabels, err := s.labelRepo.ListByKey(ctx, def.Tenant, def.Key)
+
+	validator, err := jsonschema.NewValidatorFromRawSchema(*def.Schema)
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service.go]: def.Schema can be nil -> then panic will be thrown. But if schema is empty or nil, we don't have to fetch labels from db and validate them.

File: components/director/internal/domain/labeldef/service_test.go [hunk 16]
```diff
@@ -140,19 +141,232 @@ func TestServiceList(t *testing.T) {
...
 		// THEN
 		require.EqualError(t, err, "while fetching Label Definitions: some error")
 	})
 }
 
+func TestServiceUpdate(t *testing.T) {
+	tenant := "tenant"
+	key := "firstName"
+
+	t.Run("success", func(t *testing.T) {
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service_test.go]: add test case that new schema is nil

File: components/director/internal/model/labledef_test.go [hunk 17]
```diff
@@ -63,3 +63,51 @@ func TestValidateLabelDef(t *testing.T) {
 	})
 
 }
+
```
Comment (aszecowka) [inline components/director/internal/model/labledef_test.go]: please rename this file because it contains typo

File: components/director/internal/model/labledef_test.go [hunk 18]
```diff
@@ -63,3 +63,51 @@ func TestValidateLabelDef(t *testing.T) {
...
+func TestValidateForUpdateLabelDef(t *testing.T) {
+	t.Run("valid input when schema not provided", func(t *testing.T) {
+		// GIVEN
```
Comment (aszecowka) [inline components/director/internal/model/labledef_test.go]: do we really need to have 2 methods: Validate and ValidateForUpdate()? - right now we have 2 very similar methods and we have to copy tests for them. If in Update method we fetch definition, get id for it and set ti to the model then we can probably remove ValidateForUpdate

File: components/director/internal/domain/label/repository_test.go (hunk 11 above)
Comment (kfurgol) [reply components/director/internal/domain/label/repository_test.go]: done

File: components/director/internal/domain/label/repository_test.go (hunk 13 above)
Comment (kfurgol) [reply components/director/internal/domain/label/repository_test.go]: done

File: components/director/internal/domain/label/repository_test.go [hunk 19]
```diff
@@ -511,7 +511,202 @@ func TestRepository_List(t *testing.T) {
...
+		mockConverter.On("FromEntity", inputItems[1]).Return(*expected[1], nil).Once()
+
+		repo := label.NewRepository(mockConverter)
+
+		db, dbMock := mockDatabase(t)
+		defer func() {
+			require.NoError(t, dbMock.ExpectationsWereMet())
+		}()
+
+		escapedQuery := regexp.QuoteMeta(`SELECT "id", "tenant_id", "key", "value", "app_id", "runtime_id" FROM "public"."labels" WHERE "key" = $1 AND "tenant_id" = $2`)
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/label/repository_test.go]: done

File: components/director/internal/domain/labeldef/resolver_test.go [hunk 20]
```diff
@@ -439,3 +440,129 @@ func contextThatHasTenant(expectedTenant string) interface{} {
 		return actualTenant == expectedTenant
 	})
 }
+
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/resolver_test.go]: done

File: components/director/internal/domain/labeldef/service.go [hunk 21]
```diff
@@ -2,19 +2,24 @@ package labeldef
...
 )
 
 type service struct {
 	repo       Repository
+	labelRepo  LabelRepository
 	uidService UIDService
 }
 
-func NewService(r Repository, uidService UIDService) *service {
+func NewService(r Repository, lr LabelRepository, uidService UIDService) *service {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service.go [hunk 22]
```diff
@@ -60,3 +75,45 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
 	}
 	return defs, nil
 }
+
+func (s *service) Update(ctx context.Context, def model.LabelDefinition) (model.LabelDefinition, error) {
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: It is no longer a case since now we return only error

File: components/director/internal/domain/labeldef/service.go [hunk 23]
```diff
@@ -60,3 +75,45 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+		return model.LabelDefinition{}, errors.Wrap(err, "while receiving Label Definition")
+	}
+
+	if ld == nil {
+		return model.LabelDefinition{}, fmt.Errorf("definition with %s key doesn't exist", def.Key)
+	}
+
+	ld.Schema = def.Schema
+
+	existingLabels, err := s.labelRepo.ListByKey(ctx, def.Tenant, def.Key)
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service_test.go [hunk 24]
```diff
@@ -140,19 +141,232 @@ func TestServiceList(t *testing.T) {
...
+		mockRepository := &automock.Repository{}
+		defer mockRepository.AssertExpectations(t)
+
+		mockLabelRepository := &automock.LabelRepository{}
+		defer mockLabelRepository.AssertExpectations(t)
+
+		newSchema := fixBasicSchema(t)
+
+		ld := model.LabelDefinition{
+			ID:     fixUID(),
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: done

File: components/director/internal/domain/labeldef/service.go (hunk 10 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service_test.go (hunk 12 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: done

File: components/director/internal/domain/labeldef/service.go (hunk 14 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service.go (hunk 15 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service_test.go (hunk 16 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: done

File: components/director/internal/model/labledef_test.go (hunk 17 above)
Comment (kfurgol) [reply components/director/internal/model/labledef_test.go]: done

File: components/director/internal/model/labledef_test.go (hunk 18 above)
Comment (kfurgol) [reply components/director/internal/model/labledef_test.go]: added TODO

File: components/director/internal/domain/labeldef/service.go [hunk 25]
```diff
@@ -60,3 +75,51 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
 	}
 	return defs, nil
 }
+
+func (s *service) Update(ctx context.Context, def model.LabelDefinition) error {
+	if err := def.ValidateForUpdate(); err != nil {
+		return errors.Wrap(err, "while validating Label Definition")
+	}
+
+	if def.Schema == nil {
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service.go]: let assume that I had definition with schema and then I decided to remove schema for it. In such case it won't be persisted in db. Write a test for such case.

File: components/director/internal/domain/labeldef/service_test.go (hunk 16 above)
Comment (aszecowka) [reply components/director/internal/domain/labeldef/service_test.go]: where???

File: components/director/internal/domain/labeldef/resolver.go [hunk 26]
```diff
@@ -124,3 +126,38 @@ func (r *Resolver) LabelDefinition(ctx context.Context, key string) (*graphql.La
...
+
+	err = r.srv.Update(ctx, ld)
+	if err != nil {
+		return nil, errors.Wrap(err, "while updating label definition")
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, errors.Wrap(err, "while committing transaction")
+	}
+
+	updatedLd, err := r.srv.Get(ctx, tnt, in.Key)
```
Comment (tgorgol) [inline components/director/internal/domain/labeldef/resolver.go]: This Get will always return error because the transaction is already committed at line 151.

File: components/director/pkg/jsonschema/validator.go [hunk 27]
```diff
@@ -64,11 +65,10 @@ func (v *validator) ValidateRaw(value interface{}) (bool, error) {
 		return true, nil
 	}
 
-	jsonLoader := gojsonschema.NewRawLoader(value)
-	result, err := v.schema.Validate(jsonLoader)
+	valueString, err := json.Marshal(value)
 	if err != nil {
-		return false, errors.Wrapf(err, "while validating value %+v against schema %+v", value, v.schema)
+		return false, err
```
Comment (crabtree) [inline components/director/pkg/jsonschema/validator.go]: IMO the error wrapper from the origin should be moved to the `ValidateString` func. Here the wrapping func should also be present.

File: components/director/internal/model/labeldef_test.go [hunk 28]
```diff
@@ -63,3 +63,51 @@ func TestValidateLabelDef(t *testing.T) {
...
+		// GIVEN
+		in := model.LabelDefinition{Key: "key"}
+		// WHEN
+		err := in.ValidateForUpdate()
+		// THEN
+		require.EqualError(t, err, "missing Tenant field")
+	})
+
+	t.Run("invalid schema", func(t *testing.T) {
+		// TODO
```
Comment (crabtree) [inline components/director/internal/model/labeldef_test.go]: This is on purpose? Please maybe just check the error in the test and then it does not need to be skipped.

File: components/director/internal/domain/label/repository_test.go [hunk 29]
```diff
@@ -511,7 +511,115 @@ func TestRepository_List(t *testing.T) {
...
+		mockConverter.On("FromEntity", inputItems[1]).Return(*expected[1], nil).Once()
+
+		repo := label.NewRepository(mockConverter)
+
+		db, dbMock := mockDatabase(t)
+		defer func() {
+			require.NoError(t, dbMock.ExpectationsWereMet())
+		}()
+
+		escapedQuery := regexp.QuoteMeta(`SELECT "id", "tenant_id", "key", "value", "app_id", "runtime_id"
```
Comment (crabtree) [inline components/director/internal/domain/label/repository_test.go]: Still, the formatting looks a bit weird.

File: components/director/internal/domain/labeldef/repository_test.go [hunk 30]
```diff
@@ -100,6 +100,87 @@ func TestRepositoryCreateLabelDefinition(t *testing.T) {
...
+
+	in := model.LabelDefinition{
+		ID:     labelDefID,
+		Tenant: tenantID,
+		Key:    "some-key",
+		Schema: &someSchema,
+	}
+
+	t.Run("successfully updated definition with schema", func(t *testing.T) {
+		db, dbMock := mockDatabase(t)
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/repository_test.go]: Formatting, additional new lines would make it more readable.

File: components/director/internal/domain/labeldef/resolver_test.go [hunk 31]
```diff
@@ -424,6 +426,133 @@ func TestQueryGivenLabelDefinition(t *testing.T) {
...
+		require.EqualError(t, err, "Cannot read tenant from context")
+	})
+
+	t.Run("got error on starting transaction", func(t *testing.T) {
+		// GIVEN
+		mockTransactioner := &pautomock.Transactioner{}
+		mockTransactioner.On("Begin").Return(nil, errors.New("some error"))
+		defer mockTransactioner.AssertExpectations(t)
+		ctx := persistence.SaveToContext(context.TODO(), nil)
+		ctx = tenant.SaveToContext(ctx, "tenant")
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/resolver_test.go]:
```suggestion
		ctx = tenant.SaveToContext(ctx, tnt)
```

File: components/director/internal/domain/labeldef/resolver.go (hunk 26 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/resolver.go]: done

File: components/director/internal/domain/label/repository_test.go (hunk 29 above)
Comment (kfurgol) [reply components/director/internal/domain/label/repository_test.go]: done

File: components/director/internal/domain/labeldef/repository_test.go (hunk 30 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/repository_test.go]: done

File: components/director/internal/domain/labeldef/resolver_test.go (hunk 31 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/resolver_test.go]: done

File: components/director/pkg/jsonschema/validator.go [hunk 32]
```diff
@@ -64,11 +65,10 @@ func (v *validator) ValidateRaw(value interface{}) (bool, error) {
...
 	}
 
-	return result.Valid(), nil
+	return v.ValidateString(string(valueString))
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/pkg/jsonschema/validator.go]: done

File: components/director/pkg/jsonschema/validator.go [hunk 33]
```diff
@@ -64,11 +65,10 @@ func (v *validator) ValidateRaw(value interface{}) (bool, error) {
 		return true, nil
 	}
 
-	jsonLoader := gojsonschema.NewRawLoader(value)
-	result, err := v.schema.Validate(jsonLoader)
+	valueString, err := json.Marshal(value)
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/pkg/jsonschema/validator.go]: done

File: components/director/pkg/jsonschema/validator.go (hunk 27 above)
Comment (kfurgol) [reply components/director/pkg/jsonschema/validator.go]: done

File: components/director/internal/domain/labeldef/service_test.go (hunk 16 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: `returns error if Label Definition schema for update is nil` inside `labeldef/service_test.go`

Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: `label definition schema for update is nil` inside `labeldef/service_test.go`

File: components/director/internal/domain/labeldef/service.go (hunk 25 above)
Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/model/labeldef_test.go (hunk 28 above)
Comment (kfurgol) [reply components/director/internal/model/labeldef_test.go]: test is being skipped for now due to `gojsonschema` library panics

File: components/director/internal/domain/labeldef/service.go [hunk 34]
```diff
@@ -60,3 +75,55 @@ func (s *service) List(ctx context.Context, tenant string) ([]model.LabelDefinit
...
+	}
+
+	if ld == nil {
+		return errors.Errorf("definition with %s key doesn't exist", def.Key)
+	}
+
+	ld.Schema = def.Schema
+
+	if def.Schema == nil {
+		if err := s.repo.Update(ctx, *ld); err != nil {
```
Comment (crabtree) [inline components/director/internal/domain/labeldef/service.go]: Please refactor the code that the repository Update method is called once.

Comment (kfurgol) [reply components/director/internal/domain/labeldef/service.go]: done

File: components/director/internal/domain/labeldef/service_test.go [hunk 35]
```diff
@@ -140,19 +141,267 @@ func TestServiceList(t *testing.T) {
...
+
+		mockLabelRepository.On("ListByKey", context.TODO(), tenant, key).Return(existingLabels, nil).Once()
+
+		ctx := context.TODO()
+		sut := labeldef.NewService(mockRepository, mockLabelRepository, nil)
+		// WHEN
+		err := sut.Update(ctx, in)
+		// THEN
+		require.Error(t, err)
+		require.EqualError(t, err, "label with key oldProperty is not valid against new schema for Runtime with ID foo")
```
Comment (aszecowka) [inline components/director/internal/domain/labeldef/service_test.go]: minor: add qoutes for 'oldProperty` and `foo`

Comment (kfurgol) [reply components/director/internal/domain/labeldef/service_test.go]: done

File: components/director/internal/domain/labeldef/repository.go [hunk 36]
```diff
@@ -121,3 +121,23 @@ func (r *repo) List(ctx context.Context, tenant string) ([]model.LabelDefinition
...
+	if err != nil {
+		return errors.Wrap(err, "while fetching persistence from context")
+	}
+
+	entity, err := r.conv.ToEntity(def)
+	if err != nil {
+		return errors.Wrap(err, "while creating Label Definition entity from model")
+	}
+
+	stmt := fmt.Sprintf(`UPDATE %s SET "schema" = :schema WHERE "id" = :id`, tableName)
```
[the remark this answers is not in the document]
Comment (kfurgol) [reply components/director/internal/domain/labeldef/repository.go]: since the id should exist because it's received in the service, there will be an error returned if no row was affected by the query
